frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,907 @@
+			let active_voters = solution.compact.voter_count() as u32;
+			let feasibility_weight = {
+				// defensive only: at the end of signed phase, snapshot will exits.
+				let desired_targets = Self::desired_targets().unwrap_or_default();
+				T::WeightInfo::feasibility_check(
+					voters,
+					targets,
+					active_voters,
+					desired_targets,
+				)
+			};
+			// the feasibility check itself has some weight
+			weight = weight.saturating_add(feasibility_weight);
+			match Self::feasibility_check(solution, ElectionCompute::Signed) {
+				Ok(ready_solution) => {
+					Self::finalize_signed_phase_accept_solution(
+						ready_solution,
+						&who,
+						deposit,
+						reward,
+					);
+					found_solution = true;
+
+					weight = weight
+						.saturating_add(T::WeightInfo::finalize_signed_phase_accept_solution());
+					break;
+				}
+				Err(_) => {
+					Self::finalize_signed_phase_reject_solution(&who, deposit);
+					weight = weight
+						.saturating_add(T::WeightInfo::finalize_signed_phase_reject_solution());
+				}
+			}
+		}
+
+		// Any unprocessed solution is pointless to even consider. Feasible or malicious,
+		// they didn't end up being used. Unreserve the bonds.
+		let discarded = all_submissions.len();
+		for SignedSubmission { who, deposit, .. } in all_submissions.drain() {
+			let _remaining = T::Currency::unreserve(&who, deposit);
+			weight = weight.saturating_add(T::DbWeight::get().writes(1));
+			debug_assert!(_remaining.is_zero());
+		}
+
+		debug_assert!(!SignedSubmissionIndices::<T>::exists());
+		debug_assert!(!SignedSubmissionNextIndex::<T>::exists());
+		debug_assert!(SignedSubmissionsMap::<T>::iter().next().is_none());
+
+		log!(debug, "closed signed phase, found solution? {}, discarded {}", found_solution, discarded);
+		(found_solution, weight)
+	}
+
+	/// Helper function for the case where a solution is accepted in the signed phase.
+	///
+	/// Extracted to facilitate with weight calculation.
+	///
+	/// Infallible
+	pub fn finalize_signed_phase_accept_solution(
+		ready_solution: ReadySolution<T::AccountId>,
+		who: &T::AccountId,
+		deposit: BalanceOf<T>,
+		reward: BalanceOf<T>,
+	) {
+		// write this ready solution.
+		<QueuedSolution<T>>::put(ready_solution);
+
+		// unreserve deposit.
+		let _remaining = T::Currency::unreserve(who, deposit);
+		debug_assert!(_remaining.is_zero());
+
+		// Reward.
+		let positive_imbalance = T::Currency::deposit_creating(who, reward);
+		T::RewardHandler::on_unbalanced(positive_imbalance);
+	}
+
+	/// Helper function for the case where a solution is accepted in the rejected phase.
+	///
+	/// Extracted to facilitate with weight calculation.
+	///
+	/// Infallible
+	pub fn finalize_signed_phase_reject_solution(who: &T::AccountId, deposit: BalanceOf<T>) {
+		let (negative_imbalance, _remaining) = T::Currency::slash_reserved(who, deposit);
+		debug_assert!(_remaining.is_zero());
+		T::SlashHandler::on_unbalanced(negative_imbalance);
```

Comment:
Should we use the `Slashed` event here? If not maybe we should get rid of the event

Reply:
Yes: https://github.com/paritytech/substrate/commit/bf4572c967a489b5a07d617aa8374887ec62f7a9.